xen/arch/arm/cpufreq/scpi_cpufreq.c

```diff
@@ -0,0 +1,353 @@
+/*
+ * xen/arch/arm/cpufreq/scpi_cpufreq.c
+ *
+ * SCPI based CPUFreq driver
+ *
+ * Based on Xen arch/x86/acpi/cpufreq/cpufreq.c
+ *
+ * Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
+ * Copyright (c) 2017 EPAM Systems.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <xen/types.h>
+#include <xen/delay.h>
+#include <xen/cpumask.h>
+#include <xen/sched.h>
+#include <xen/xmalloc.h>
+#include <xen/err.h>
+#include <xen/cpufreq.h>
+#include <asm/bug.h>
+#include <asm/percpu.h>
+
+#include "scpi_protocol.h"
+
+extern struct device *get_cpu_device(unsigned int cpu);
+
+struct scpi_cpufreq_data
+{
+    struct processor_performance *perf;
+    struct cpufreq_frequency_table *freq_table;
+    struct scpi_dvfs_info *info;
+    int domain_id;
+};
+
+static struct scpi_cpufreq_data *cpufreq_driver_data[NR_CPUS];
+
+static struct cpufreq_driver scpi_cpufreq_driver;
+
+static struct scpi_ops *scpi_ops;
+
+static unsigned int scpi_cpufreq_get(unsigned int cpu)
+{
+    struct scpi_cpufreq_data *data;
+    struct cpufreq_policy *policy;
+    const struct scpi_opp *opp;
+    int idx;
+
+    if ( cpu >= nr_cpu_ids || !cpu_online(cpu) )
+        return 0;
+
+    policy = per_cpu(cpufreq_cpu_policy, cpu);
+    if ( !policy || !( data = cpufreq_driver_data[policy->cpu] ) )
+        return 0;
+
+    /* Sanity check */
+    if ( data->domain_id < 0 || !data->info )
+        return 0;
+
+    idx = scpi_ops->dvfs_get_idx(data->domain_id);
+    if ( idx < 0 )
+        return 0;
+
+    opp = data->info->opps + idx;
+
+    return opp->freq / 1000; /* kHz */
+}
+
+static unsigned int check_freq(unsigned int cpu, unsigned int freq)
+{
+    unsigned int curr_freq;
+    int i;
+
+    for ( i = 0; i < 100; i++ )
+    {
+        curr_freq = scpi_cpufreq_get(cpu);
+        if ( curr_freq == freq )
+            return 1;
+        udelay(10);
```

Comment:
10 microsends * 100 = 1 millisecond. Isn't this too much?

Reply:
On the whole it is a long period of time, but it depends... Please see, my explanation above

Reply:
We decided to drop this func.